pkg/apis/batch/validation/validation.go

```diff
@@ -179,10 +196,99 @@ func validateJobSpec(spec *batch.JobSpec, fldPath *field.Path, opts apivalidatio
+			if rule.OnExitCodes.Operator == "" {
+				operatorPath := onExitCodesPath.Child("operator")
+				allErrs = append(allErrs, field.Required(operatorPath, fmt.Sprintf("valid values: %q", supportedPodFailurePolicyOnExitCodesOperator.List())))
+			} else if !supportedPodFailurePolicyOnExitCodesOperator.Has(string(rule.OnExitCodes.Operator)) {
+				path := onExitCodesPath.Child("operator")
+				allErrs = append(allErrs, field.NotSupported(path, rule.OnExitCodes.Operator, supportedPodFailurePolicyOnExitCodesOperator.List()))
+			}
+			if rule.OnExitCodes.ContainerName != nil && !containerNames.Has(*rule.OnExitCodes.ContainerName) {
+				path := onExitCodesPath.Child("containerName")
+				allErrs = append(allErrs, field.Invalid(path, *rule.OnExitCodes.ContainerName, "must be one of the container names in the pod template"))
+			}
+			valuesPath := onExitCodesPath.Child("values")
+			if len(rule.OnExitCodes.Values) == 0 {
+				allErrs = append(allErrs, field.Invalid(valuesPath, rule.OnExitCodes.Values, "at least one value is required"))
+			} else {
+				isOrdered := true
+				uniqueValues := sets.NewInt32()
+				for j, exitCodeValue := range rule.OnExitCodes.Values {
+					if rule.OnExitCodes.Operator == batch.PodFailurePolicyOnExitCodesOpIn && exitCodeValue == 0 {
+						valuePath := valuesPath.Index(j)
+						allErrs = append(allErrs, field.Invalid(valuePath, exitCodeValue, "must not be 0 for the In operator"))
+					}
+					uniqueValues.Insert(rule.OnExitCodes.Values[j])
+					if j > 0 && rule.OnExitCodes.Values[j-1] > rule.OnExitCodes.Values[j] {
+						isOrdered = false
+					}
+				}
+				if uniqueValues.Len() < len(rule.OnExitCodes.Values) {
+					allErrs = append(allErrs, field.Invalid(valuesPath, rule.OnExitCodes.Values, "must not contain duplicates"))
+				}
+				if !isOrdered {
+					allErrs = append(allErrs, field.Invalid(valuesPath, rule.OnExitCodes.Values, "must be ordered"))
+				}
+			}
+		}
+		if rule.OnPodConditions != nil {
+			onPodConditionsPath := rulePath.Child("onPodConditions")
+			for j, pattern := range rule.OnPodConditions {
+				patternPath := onPodConditionsPath.Index(j)
+				allErrs = append(allErrs, apivalidation.ValidateQualifiedName(string(pattern.Type), patternPath.Child("type"))...)
+			}
+		}
+		if rule.OnExitCodes != nil && rule.OnPodConditions != nil {
+			if ruleStr := podFailurePolicyRuleToString(rule); ruleStr != nil {
+				allErrs = append(allErrs, field.Invalid(rulePath, ruleStr, "Specifying both OnExitCodes and OnPodConditions is not supported"))
+			}
+		}
+		if rule.OnExitCodes == nil && rule.OnPodConditions == nil {
+			if ruleStr := podFailurePolicyRuleToString(rule); ruleStr != nil {
+				allErrs = append(allErrs, field.Invalid(rulePath, ruleStr, "Specifying one of OnExitCodes and OnPodConditions is required"))
+			}
+		}
 	}
 	return allErrs
 }
 
+func podFailurePolicyRuleToString(rule batch.PodFailurePolicyRule) *string {
+	ruleJsonBytes, err := json.Marshal(rule)
```

Comment:
I see, this validation has already some history, so let me summarize :).

The issue is that the `field.Invalid()` function needs to get the value as the second parameter and it serializes it to string. The default serialization puts the pointer memory locations into string, which feels unsafe from the security perspective. Also, it would make the test harder to write, cause the error message could differ between runs (due to differing mem locations).

The solutions discussed so far:
1. use json Marshal as currently (but it has the issue of silencing errors, cause the validation methods don't propagate them)
2. describe the value (say `[pod failure policy rule at index 1]`) - initial implementation, but review comments suggested to explore other options,
3. use `Forbidden` error which does not require value - the error type is misleading (already raised as an issue in another comment, that Invalid is the preferred type)
4. create two Invalid errors (one per OnExitCodes and OnPodConditions) - does not really solve anything because the default serialization for OnExitCodes would also contain the pointer for the containerName field
5. accept that we return the memory pointers and make the test asserts more flexible -  I don't have enough expertise to say if this is safe, but it seems very risky to me. Also, it feels to me like running into work to adjust many subtests.

My personal preference is 2. simple, safe and human readable. Your opinions / preferences / other ideas @alculquicondor @soltysh @liggitt ?

Reply:
I'm fine with option 1.
option 2 is just repetitive, because we already have the path in the message.

Yet another is being to use Invalid while saying that the value is not important and doesn't need to be printed. This should be possible with a special type. And then you could call it like this:

```golang
field.Invalid(path, field.Redacted{}, msg)
```

But I would leave that for a follow up.

Reply:
Ok, it turned out the idea with Redacted is very simple to implement with the use of field.OmitValueType{} - just had to export it, hope it is not a deal.